source/layers/validation/checkers/events_deadlock/zel_events_deadlock_checker.cpp

```diff
@@ -0,0 +1,587 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * @file zel_events_deadlock_checker.cpp
+ *
+ */
+#include "zel_events_deadlock_checker.h"
+
+#include <iostream>
+#include <sstream>
+
+namespace validation_layer {
+class eventsDeadlockChecker eventsDeadlock_checker;
+
+eventsDeadlockChecker::eventsDeadlockChecker() {
+
+    enableEventsDeadlock = getenv_tobool("ZEL_ENABLE_EVENTSDEADLOCK_CHECKER");
+    if (enableEventsDeadlock) {
+        eventsDeadlockChecker::ZEeventsDeadlockChecker *zeChecker = new eventsDeadlockChecker::ZEeventsDeadlockChecker;
+        eventsDeadlockChecker::ZESeventsDeadlockChecker *zesChecker = new eventsDeadlockChecker::ZESeventsDeadlockChecker;
+        eventsDeadlockChecker::ZETeventsDeadlockChecker *zetChecker = new eventsDeadlockChecker::ZETeventsDeadlockChecker;
+        eventsDeadlock_checker.zeValidation = zeChecker;
+        eventsDeadlock_checker.zesValidation = zesChecker;
+        eventsDeadlock_checker.zetValidation = zetChecker;
+
+        validation_layer::context.validationHandlers.push_back(&eventsDeadlock_checker);
+    }
+}
+
+eventsDeadlockChecker::~eventsDeadlockChecker() {
+    if (enableEventsDeadlock) {
+        delete eventsDeadlock_checker.zeValidation;
+    }
+}
+
+ze_result_t eventsDeadlockChecker::ZEeventsDeadlockChecker::zeEventCreateEpilogue(
+    ze_event_pool_handle_t hEventPool, ///< [in] handle of the event pool
+    const ze_event_desc_t *desc,       ///< [in] pointer to event descriptor
+    ze_event_handle_t *phEvent         ///< [out] pointer to handle of event object created
+) {
+    eventToDagID[*phEvent] = invalidDagID;
+
+    return ZE_RESULT_SUCCESS;
+}
+
+ze_result_t
+eventsDeadlockChecker::ZEeventsDeadlockChecker::zeEventDestroyEpilogue(
+    ze_event_handle_t hEvent ///< [in][release] handle of event object to destroy
+) {
+    if (eventToDagID.find(hEvent) != eventToDagID.end()) {
+        // Delete event from eventToDagID but not from the dagIDToAction map as it may be needed for printing the discription of the action when printing path in the DAG.
+        eventToDagID.erase(hEvent);
+    }
+
+    return ZE_RESULT_SUCCESS;
```

Comment:
general question: how is epilogue call handled? if the call to driver returns an error then does epilogue get called? also does epilogue need to return the return value from the driver call?

Reply:
@nrspruit do you know?

Reply:
regardless of the result, the epilogue will be called like here:
https://github.com/oneapi-src/level-zero/blob/master/source/layers/validation/ze_valddi.cpp#L49 .

Perhaps we could expand this to pass in the "result" to the epilogue such that a check does not occur or don't call epilogue given an error?

Reply:
Yes, either not call epilogue altogether if error in the actual zeCall or pass the error in the epilogue and then its as per the checker logic and behavior what it needs to do in case of error. 

In the later case, then it will have to change API for all checkers for all zeAPI calls. Or just for the ZeEventDestroy()? 

Again, in the former case it will be only for ZeEventDestroy() or all ZeAPI calls?